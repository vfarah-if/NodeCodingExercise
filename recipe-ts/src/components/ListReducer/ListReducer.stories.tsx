import React from 'react'
import { Story, Meta } from '@storybook/react/types-6-0'
import ListReducer, { ListReducerProps } from './ListReducer'

export default {
  title: 'Component/ListReducer',
  component: ListReducer,
} as Meta

export const DefaultListReducer: Story<ListReducerProps> = () => (
  <>
    <ListReducer />
  </>
)
