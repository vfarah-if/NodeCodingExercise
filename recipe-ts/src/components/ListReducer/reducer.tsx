export type ACTIONTYPE =
  | { type: 'newItemChange'; payload: string }
  | { type: 'add' }

export const initialState = {
  newItem: '',
  items: new Array<string>(),
}

export const reducer = (state: typeof initialState, action: ACTIONTYPE) => {
  switch (action.type) {
    case 'newItemChange':
      return {
        ...state,
        newItem: action.payload,
      }
    case 'add':
      return {
        items: state.items.concat([state.newItem]),
        newItem: '',
      }
  }
}
