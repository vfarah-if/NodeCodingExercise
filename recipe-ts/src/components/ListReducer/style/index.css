.todo-list {
	margin: 20px;
    padding: 20px;
	border: solid medium lightgrey;
	border-radius: 10px;
    background-color: gray;
}

.form-group {
	margin-bottom: 1rem;
}

.form-control {
	display: block;
	width: 100%;
	padding: 0.375rem;
	font-size: 1rem;
	line-height: 1.5;
	color: #495057;
	background-color: #fff;
	background-clip: padding-box;
	border: 1px solid #ced4da;
	border-radius: 0.25rem;
	transition: border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
}

btn {
	display: inline-block;
	font-weight: 400;
	text-align: center;
	white-space: nowrap;
	vertical-align: middle;
	-webkit-user-select: none;
	-moz-user-select: none;
	-ms-user-select: none;
	user-select: none;
	border: 1px solid transparent;
	padding: 0.375rem 0.75rem;
	font-size: 1rem;
	line-height: 1.5;
	border-radius: 0.25rem;
	transition: color 0.15s ease-in-out, background-color 0.15s ease-in-out,
		border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
}

.btn:hover,
.btn:focus {
	text-decoration: none;
}

.btn:focus,
.btn.focus {
	outline: 0;
	box-shadow: 0 0 0 0.2rem rgba(0, 123, 255, 0.25);
}

.btn.disabled,
.btn:disabled {
	opacity: 0.65;
}

.btn:not(:disabled):not(.disabled) {
	cursor: pointer;
}

.btn:not(:disabled):not(.disabled):active,
.btn:not(:disabled):not(.disabled).active {
	background-image: none;
}

.btn-primary {
	color: #fff;
	background-color: #007bff;
	border-color: #007bff;
}

.btn-primary:hover {
	color: #fff;
	background-color: #0069d9;
	border-color: #0062cc;
}

.btn-primary:focus,
.btn-primary.focus {
	box-shadow: 0 0 0 0.2rem rgba(0, 123, 255, 0.5);
}

.btn-primary.disabled,
.btn-primary:disabled {
	color: #fff;
	background-color: #007bff;
	border-color: #007bff;
}

.btn-primary:not(:disabled):not(.disabled):active,
.btn-primary:not(:disabled):not(.disabled).active,
.show > .btn-primary.dropdown-toggle {
	color: #fff;
	background-color: #0062cc;
	border-color: #005cbf;
}

.btn-primary:not(:disabled):not(.disabled):active:focus,
.btn-primary:not(:disabled):not(.disabled).active:focus,
.show > .btn-primary.dropdown-toggle:focus {
	box-shadow: 0 0 0 0.2rem rgba(0, 123, 255, 0.5);
}
