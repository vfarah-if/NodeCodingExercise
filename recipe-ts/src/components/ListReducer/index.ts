
import ListReducer from './ListReducer';
export default ListReducer;

export { reducer } from './reducer';