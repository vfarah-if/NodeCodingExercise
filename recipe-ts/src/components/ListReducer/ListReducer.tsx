import React, { useReducer } from 'react';
import { initialState, reducer } from './reducer';
import './style/index.css';

export interface ListReducerProps {}


const ListReducer: React.FC<ListReducerProps> = () => {
	const [state, dispatch] = useReducer(reducer, initialState);

	const submitHandler = (event: React.FormEvent<HTMLFormElement>) => {
		event.preventDefault();
		dispatch({ type: 'add' });
	};

	return (
		<div className="todo-list">
			<h2>TODO List</h2>
			<ul>
				{state.items.map((item) => (
					<li key={item}>{item}</li>
				))}
			</ul>
			<form onSubmit={submitHandler} data-testid="form">
				<div className="form-group">
					<input
						id="newItem"
						className="form-control"
						type="text"
						value={state.newItem}
						onChange={(e) =>
							dispatch({
								type: 'newItemChange',
								payload: e.target.value,
							})
						}
						placeholder="Enter new item"
					/>
				</div>
				<input
					type="submit"
					value="Add"
					className="btn btn-primary"
				></input>
			</form>
		</div>
	);
};

export default ListReducer;
