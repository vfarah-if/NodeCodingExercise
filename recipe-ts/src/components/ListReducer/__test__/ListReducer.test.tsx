import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import ListReducer from '../ListReducer';

describe('List', () => {
	const item1 = 'TODO 1: Add the first item';
	const item2 = 'TODO 2: Add the second item';

	test('should add one item into the list', async () => {
		const component = render(<ListReducer />);
		const input = (await component.findByPlaceholderText(
			'Enter new item'
		)) as HTMLInputElement;
		editValue(input, item1);

		await submitAddButton(component);

		const listItem = await component.findByText(item1);
		expect(listItem).toMatchSnapshot();
	});

	test('should clear the list after adding another item', async () => {
		const component = render(<ListReducer />);
		const input = (await component.findByPlaceholderText(
			'Enter new item'
		)) as HTMLInputElement;

		editValue(input, item1);
		await submitAddButton(component);
		expect(input.value).toBe('');

		editValue(input, item2);
		await submitAddButton(component);
		expect(input.value).toBe('');

		const listItem = await component.findByText(item2);
        expect(listItem).toMatchSnapshot();
        expect(component.container).toMatchSnapshot('both items')
	});
});

async function submitAddButton(component: any) {
	const form = await component.findByTestId('form');
	expect(form).toBeDefined();
	fireEvent.submit(form);
}

function editValue(input: HTMLInputElement, value: string) {
	expect(input).toBeDefined();
	fireEvent.change(input, {
		target: { value },
	});
	expect(input.value).toBe(value);
}
