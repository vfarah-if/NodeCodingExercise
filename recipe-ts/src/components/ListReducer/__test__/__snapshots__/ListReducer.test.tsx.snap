// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should add one item into the list 1`] = `
<li>
  TODO 1: Add the first item
</li>
`;

exports[`List should clear the list after adding another item 1`] = `
<li>
  TODO 2: Add the second item
</li>
`;

exports[`List should clear the list after adding another item: both items 1`] = `
<div>
  <div
    class="todo-list"
  >
    <h2>
      TODO List
    </h2>
    <ul>
      <li>
        TODO 1: Add the first item
      </li>
      <li>
        TODO 2: Add the second item
      </li>
    </ul>
    <form
      data-testid="form"
    >
      <div
        class="form-group"
      >
        <input
          class="form-control"
          id="newItem"
          placeholder="Enter new item"
          type="text"
          value=""
        />
      </div>
      <input
        class="btn btn-primary"
        type="submit"
        value="Add"
      />
    </form>
  </div>
</div>
`;
